File: frontend/Ecommerce/src/app/product.ts
export interface Product {
  productName: string;
  price: number;
  quantity: number;
  description: string;
  categoryBrandPairs: {
    categoryId: number;
    brandId: number | null;
  }[];
}

export interface ProductList{
  id: number;
  productName: string;
  productCode: string;
  description: string;
  price: number;
  quantity: number;
  createDate : string;
  updateDate : string;
  status : number;
  productImages: {
    id: number;
    imageUrl: string;
    status : number;
  }[];
  checked?: boolean;
}

export interface ProductDTO {
  id: number;
  productName: string;
  productCode: string;
  price: number;
  quantity: number;
  description: string;
  status: number;

  categoryBrandArray: {
    categoryId: number;
    cateName?: string;
    brandId: number | null;
    brandName?: string | null;
  }[];
  
  productImages: {
    id: number;
    imageUrl: string;
    status: number;
  }[];
  
  // Rating fields
  averageRating?: number;
  reviewCount?: number;
}

File: backend/Ecommerce/src/main/java/com/Ojt/Ecommerce/entity/Product.java
package com.Ojt.Ecommerce.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import jakarta.persistence.*;
import lombok.*;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Entity
@Table(name = "products")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "product_name", length = 45)
    private String productName;

    private Double price;

    @Column(name = "product_code", length = 45)
    private String productCode;

    @Column(name = "status", columnDefinition = "INT DEFAULT 1")
    private Integer status;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(name = "create_date")
    private LocalDateTime createDate;

    @Column(name = "update_date")
    private LocalDateTime updateDate;

    @Column(name = "quantity")
    private Long quantity;

    @ManyToOne
    @JoinColumn(name = "brand_id")
    private Brand brand;

//    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
//    @JsonIgnore
//    private List<ProductHasCategory> productCategories;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL,fetch = FetchType.EAGER, orphanRemoval = true)
    @JsonIgnore
    @ToString.Exclude //add this two line to fix infinite loop error
    @EqualsAndHashCode.Exclude
    private Set<ProductHasCategory> productCategories; // Changed from List to Set

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL,fetch = FetchType.EAGER, orphanRemoval = true)
    private List<ProductImage> productImages; // Keep this as List

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonIgnore
    private List<Wishlist> wishlists;

    @OneToMany(mappedBy = "product")
    @JsonIgnore
    private List<UserOrderHasProduct> orderProducts;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonIgnore
    private List<ProductVariant> productVariants;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<PurchaseProducts> purchaseProducts;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<ProductDiscount> productDiscounts;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonIgnore
    private List<Review> reviews = new ArrayList<>();

    @PrePersist
    public void prePersist() {
        if (status == null) {
            status = 1;
        }
    }
}